subprojects/base-services/src/main/java/org/gradle/internal/reflect/ReflectionCache.java

```diff
@@ -34,12 +34,14 @@ public T get(Class<?> key) {
         synchronized (lock) {
             cached = cache.get(key);
         }
-        if (cached != null) {
+        if (cached != null && !hasExpired(cached)) {
```

Comment:
It is probably possible. I thought myself the same, but didn't manage to put the VM in such a stress that it highlighted this behavior (as opposed to the previous one which was definitely visible). I'm still trying to figure out if there's a better way to do this. Use of `ClassValue` under JDK 8 would be possible and safer I guess, but it would involve maintaining 2 different caches.